File: src/Stats.LogInterpretation/ClientNameTranslation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Stats.LogInterpretation
{
    public static class ClientNameTranslation
    {
        public static string GetClientCategory(string clientName)
        {
            if (string.IsNullOrWhiteSpace(clientName))
            {
                return string.Empty;
            }

            if (clientName.ContainsAnyClientName(ClientNames.NuGet))
            {
                return ClientCategories.NuGet;
            }

            if (clientName.ContainsAnyClientName(ClientNames.WebMatrix))
            {
                return ClientCategories.WebMatrix;
            }

            if (clientName.ContainsAnyClientName(ClientNames.NuGetPackageExplorer))
            {
                return ClientCategories.NuGetPackageExplorer;
            }

            if (clientName.ContainsAnyClientName(ClientNames.Script))
            {
                return ClientCategories.Script;
            }

            if (clientName.ContainsAnyClientName(ClientNames.Crawler))
            {
                return ClientCategories.Crawler;
            }

            if (clientName.ContainsAnyClientName(ClientNames.Mobile))
            {
                return ClientCategories.Mobile;
            }

            // Check these late in the process, because other User Agents tend to also
            // send browser strings (e.g. PowerShell sends the Mozilla string along).
            if (clientName.ContainsAnyClientName(ClientNames.Browser)
                || ClientNames.AbsoluteBrowserNames
                    .Any(abn => clientName.Equals(abn, StringComparison.OrdinalIgnoreCase)))
            {
                return ClientCategories.Browser;
            }

            // Explicitly categorize unknowns, test frameworks or others that should be filtered out in the reports
            if (clientName.ContainsAnyClientName(ClientNames.Unknown))
            {
                return ClientCategories.Unknown;
            }

            // Return empty for all others to allow ecosystem user agents
            // to be picked up in the reports
            return string.Empty;
        }

        /// <summary>
        /// Extension method for ignore case `Contains` check.
        /// </summary>
        /// <param name="source">Source string in which to check the target</param>
        /// <param name="target">Target substring to verify</param>
        /// <param name="comparison">Instance of <see cref="StringComparison"/></param>
        /// <returns></returns>
        private static bool Contains(this string source, string target, StringComparison comparison)
        {
            return source?.IndexOf(target, comparison) >= 0;
        }

        /// <summary>
        /// Extension method that checks if any of the strings in the target list are a substring 
        /// in the source string.
        /// </summary>
        /// <param name="source">String to be checked</param>
        /// <param name="targetList">List of substrings to verify</param>
        /// <param name="comparison">Instane of <see cref="StringComparison"/>, defaults to <see cref="StringComparison.OrdinalIgnoreCase"/></param>
        /// <returns></returns>
        private static bool ContainsAnyClientName(this string source, IReadOnlyList<string> targetList, StringComparison comparison = StringComparison.OrdinalIgnoreCase)
        {
            if (string.IsNullOrWhiteSpace(source))
            {
                return false;
            }

            return targetList.Any(t => source.Contains(t, comparison));
        }
    }
}

File: python/StatsLogParser/loginterpretation/clientnametranslation.py
from dataclasses import dataclass, field
from typing import List

# Client Categories using dataclass
@dataclass(frozen=True)
class ClientCategories:
    nuget: str = "NuGet"
    webmatrix: str = "WebMatrix"
    nuget_package_explorer: str = "NuGet Package Explorer"
    script: str = "Script"
    crawler: str = "Crawler"
    mobile: str = "Mobile"
    browser: str = "Browser"
    unknown: str = "Unknown"

# Client Names using dataclass
@dataclass(frozen=True)
class ClientNames:
    nuget: List[str] = field(default_factory=lambda: [
        "NuGet Cross-Platform Command Line",
        "NuGet Client V3",
        "NuGet VS VSIX",
        "NuGet VS PowerShell Console",
        "NuGet VS Packages Dialog - Solution",
        "NuGet VS Packages Dialog",
        "NuGet Shim",
        "NuGet Add Package Dialog",
        "NuGet Command Line",
        "NuGet Package Manager Console",
        "NuGet Visual Studio Extension",
        "Package-Installer",
        "NuGet MSBuild Task",
        "NuGet .NET Core MSBuild Task",
        "NuGet Desktop MSBuild Task"
    ])

    webmatrix: List[str] = field(default_factory=lambda: ["WebMatrix"])

    nuget_package_explorer: List[str] = field(default_factory=lambda: [
        "NuGet Package Explorer Metro",
        "NuGet Package Explorer"
    ])

    script: List[str] = field(default_factory=lambda: [
        "Powershell",
        "curl",
        "Wget",
        "Java"
    ])

    crawler: List[str] = field(default_factory=lambda: [
        "Bot",
        "bot",
        "Slurp",
        "BingPreview",
        "crawler",
        "sniffer",
        "spider"
    ])

    mobile: List[str] = field(default_factory=lambda: [
        "Mobile",
        "Android",
        "Kindle",
        "BlackBerry",
        "Openwave",
        "NetFront",
        "CFNetwork",
        "iLunascape"
    ])

    browser: List[str] = field(default_factory=lambda: [
        "Mozilla",
        "Firefox",
        "Opera",
        "Chrome",
        "Chromium",
        "Internet Explorer",
        "Browser",
        "Safari",
        "Sogou Explorer",
        "Maxthon",
        "SeaMonkey",
        "Iceweasel",
        "Sleipnir",
        "Konqueror",
        "Lynx",
        "Galeon",
        "Epiphany",
        "Lunascape"
    ])

    absolute_browser_names: List[str] = field(default_factory=lambda: [
        "IE",
        "Iron"
    ])

    unknown: List[str] = field(default_factory=lambda: [
        "PhantomJS",
        "WebKit Nightly",
        "Python Requests",
        "Jasmine",
        "Java",
        "AppleMail",
        "NuGet Test Client"
    ])

# Client Name Translation Logic using static methods
class ClientNameTranslation:
    @staticmethod
    def get_client_category(client_name: str) -> str:
        if not client_name or client_name.strip() == "":
            return ""

        if ClientNameTranslation.contains_any_client_name(client_name, ClientNames().nuget):
            return ClientCategories().nuget

        if ClientNameTranslation.contains_any_client_name(client_name, ClientNames().webmatrix):
            return ClientCategories().webmatrix

        if ClientNameTranslation.contains_any_client_name(client_name, ClientNames().nuget_package_explorer):
            return ClientCategories().nuget_package_explorer

        if ClientNameTranslation.contains_any_client_name(client_name, ClientNames().script):
            return ClientCategories().script

        if ClientNameTranslation.contains_any_client_name(client_name, ClientNames().crawler):
            return ClientCategories().crawler

        if ClientNameTranslation.contains_any_client_name(client_name, ClientNames().mobile):
            return ClientCategories().mobile

        # Check these late in the process, because other User Agents tend to also send browser strings
        if (ClientNameTranslation.contains_any_client_name(client_name, ClientNames().browser) or
            any(client_name.strip().lower() == abn.lower() for abn in ClientNames().absolute_browser_names)):
            return ClientCategories().browser

        # Explicitly categorize unknowns, test frameworks, or others that should be filtered out in the reports
        if ClientNameTranslation.contains_any_client_name(client_name, ClientNames().unknown):
            return ClientCategories().unknown

        # Return empty for all others to allow ecosystem user agents to be picked up in the reports
        return ""

    @staticmethod
    def contains(source: str, target: str, comparison=str.casefold) -> bool:
        return comparison(target) in comparison(source)

    @staticmethod
    def contains_any_client_name(source: str, target_list: List[str], comparison=str.casefold) -> bool:
        if not source or source.strip() == "":
            return False
        return any(ClientNameTranslation.contains(source, target, comparison) for target in target_list)
